### PR TITLE
SpeechLLM LibriSpeech recipe

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

Not ready to be merged yet :) 

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/integrations/huggingface/llama.py [hunk 1]
```diff
@@ -63,13 +66,22 @@ def __init__(
...
         self.source = source
         self.save_path = save_path
         self.bnb_config = bnb_config
 
+        # Capture config-only overrides to avoid passing them to from_pretrained
+        self._config_overrides = {}
+        if "output_hidden_states" in kwargs:
+            self._config_overrides["output_hidden_states"] = kwargs.pop(
+                "output_hidden_states"
+            )
```
Comment (Adel-Moumen) [inline speechbrain/integrations/huggingface/llama.py:87]: TBH, I don't remember in what scenario you want to set `output_hidden_states=True`

File: recipes/LibriSpeech/ASR/transformer/train_speechllm.py [hunk 2]
```diff
@@ -0,0 +1,678 @@
...
+        """
+        yield wrd
+        tokens_list = tokenizer(wrd, add_special_tokens=False).input_ids
+        yield tokens_list
+        tokens_bos = torch.LongTensor(
+            [start_of_audio_index]
+            + [end_of_audio_index]
+            + prompt_ids
+            + [bos_index]
+            + tokens_list
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/transformer/train_speechllm.py:475]: one thing: it's really hard to know if an LM requires bos / eos (e.g. https://huggingface.co/blog/qgallouedec/gotchas-in-tokenizer-behavior) ; so, I think, ideally, we need to have a proper prompt function that will create a prompt depending on the input tokens (e.g. is eos is None, then don't append etc).

File: recipes/LibriSpeech/ASR/transformer/hparams/speechllm_ssl_feats.yaml
Comment (pplantinga) [inline recipes/LibriSpeech/ASR/transformer/hparams/speechllm_ssl_feats.yaml:1]: Do we need a second yaml file for this? I'm legitimately curious -- its fine if the answer is yes!

File: speechbrain/integrations/huggingface/llama.py [hunk 3]
```diff
@@ -63,13 +66,22 @@ def __init__(
         freeze: bool = False,
         pad_token: str = "[PAD]",
         torch_dtype: torch.dtype = torch.float16,
+        additional_special_tokens: List[str] = None,
+        pad_to_multiple_of: int = 8,
```
Comment (pplantinga) [inline speechbrain/integrations/huggingface/llama.py:74]: These new parameters are not documented.

File: speechbrain/nnet/activations.py
Comment (pplantinga) [inline speechbrain/nnet/activations.py:1]: What's the benefit of adding this over `torch.nn.GELU`? It doesn't seem to behave any differently

File: recipes/LibriSpeech/ASR/transformer/hparams/speechllm_e2e.yaml [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                value: !ref <ignore_index>
+
+####################### Model Architecture Parameters #######################
+lora_rank: 16
+
+# Output & Decoding Parameters
+bos_index: !PLACEHOLDER # 0
+eos_index: !PLACEHOLDER # 0
+pad_token: !PLACEHOLDER # 128256
+prompt: "Transcribe speech to text."
```
Comment (pplantinga) [inline recipes/LibriSpeech/ASR/transformer/hparams/speechllm_e2e.yaml:140]: This recipe doesn't support tasks other than transcription? Couldn't we at least do keyword spotting? "Is the word {word} present in the audio?" For SpeechLLMs my understanding is that we ultimately want multi-task machines, so it would be nice if we at least had a basic concept of how multi-task would be handled.

**Review by pplantinga — COMMENTED** [review]
Overall, looks like a good addition -- it will be nice to have a starting framework for doing SpeechLLMs in SpeechBrain!

From my first read, I guess I'm wondering if we want to try to support more than one task, even a simple second task such as keyword spotting, just to show how it can be done, as it seems like the main benefit of SpeechLLMs over traditional ASR is the fact that they can support multiple tasks.

File: recipes/LibriSpeech/ASR/transformer/README.md [hunk 5]
```diff
@@ -23,6 +22,20 @@ installed in your environment (see extra-requirements.txt)**
...
+- SpeechLLM with SSL features
+- SpeechLLM with E2E features
+
+In the first mode, the speech features are extracted from the audio waveforms using a pre-trained SSL model, and then projected to the LLM embedding space using a linear layer projection, where everything is trained jointly.
+
+In the second mode, the speech features are already being extracted offline (see: `extract_ssl_feats.py` script). The LLM is then trained on the frozen SSL representations. This mode is more efficient and faster to train, but at the cost of flexibility on the frozen SSL model.
+
+| Release | Model | hyperparams file | Dev Clean WER | Test Clean WER | Test Other WER | HuggingFace link | Model link | GPUs |
+|:-------------:|:-------------:|:-------------:|:---------------------------:| :-----:| :-----:| :-----:| :--------:|
+| 29-11-25 | WavLM Large + SmolLM2 1.7B + LoRA | speechllm_ssl_feats.yaml | N/A | 3.17 | 6.83 | [HuggingFace](https://huggingface.co/speechbrain/asr-speechllm-librispeech) | - | 1xH100 40GB |
```
Comment (TParcollet) [inline recipes/LibriSpeech/ASR/transformer/README.md]: These are not really good numbers given the architecture imho, but ok.

Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/README.md]: I had much better results (2.X%), but I need to train the models a bit longer.

File: recipes/LibriSpeech/ASR/transformer/extract_ssl_feats.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+"""Script to extract SSL features from the audio waveforms.
+
+The script uses the `speechbrain.integrations.hdf5.cached_item` module to cache the features.
```
Comment (TParcollet) [inline recipes/LibriSpeech/ASR/transformer/extract_ssl_feats.py:4]: Tutorial

File: recipes/LibriSpeech/ASR/transformer/hparams/speechllm_e2e.yaml (hunk 4 above)
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/hparams/speechllm_e2e.yaml:140]: I think for now it would best to stay on a single task, since we already have a speechllm translation task (although, will need some udpates in the future as the current code is not optimal). One could just train LORA adapters for ASR, then again a new set of adapters for a different task while loading the same backbone (instead of training with multiple tasks which I think what is currently being done in industry)

File: recipes/LibriSpeech/ASR/transformer/hparams/speechllm_ssl_feats.yaml
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/hparams/speechllm_ssl_feats.yaml:1]: I think it's better to have two yamls so that we don't end up loading on the device the full architecture while we should only load the encoder for extracting features.

File: recipes/LibriSpeech/ASR/transformer/extract_ssl_feats.py (hunk 6 above)
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/extract_ssl_feats.py:4]: Done! Please have a look at data-loading-pipeline which now features multiple sections about saving features, and also update the speaker example with loading saved features.

File: recipes/LibriSpeech/ASR/transformer/README.md [hunk 7]
```diff
@@ -23,6 +22,20 @@ installed in your environment (see extra-requirements.txt)**
...
+- SpeechLLM with SSL features
+- SpeechLLM with E2E features
+
+In the first mode, the speech features are extracted from the audio waveforms using a pre-trained SSL model, and then projected to the LLM embedding space using a linear layer projection, where everything is trained jointly.
+
+In the second mode, the speech features are already being extracted offline (see: `extract_ssl_feats.py` script). The LLM is then trained on the frozen SSL representations. This mode is more efficient and faster to train, but at the cost of flexibility on the frozen SSL model.
+
+| Release | Model | hyperparams file | Dev Clean WER | Dev Other WER | Test Clean WER | Test Other WER | HuggingFace link | Model link | GPUs |
+|:-------------:|:-------------:|:-------------:|:---------------------------:| :-----:| :-----:| :-----:| :-----:| :--------:|
+| 29-01-26 | WavLM Large + LLama 3.2 1B + LoRA | speechllm_e2e.yaml | 2.79 | 5.03 | 2.72 | 5.34 | [HuggingFace](https://huggingface.co/speechbrain/asr-wavlm-large-llama3.2-1b-lora-librispeech) | - | 1xA100 80GB |
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/transformer/README.md:37]: results have been updated and now are much better! Will upload and turn the repo public ASAP.

File: speechbrain/nnet/activations.py
Comment (Adel-Moumen) [reply speechbrain/nnet/activations.py:1]: The main reason is standardisation with the yamls. I found in some examples that we had to define a class in order to use it in speechbrain e.g. if this an apply function (i.e. return f(x)) or a standalone class that you need first to create (e.g. f = nn.ReLU(); f(x))

Comment (Adel-Moumen) [reply speechbrain/nnet/activations.py:1]: I don't think this is necessary a bad idea, especially when you think about the fact that some classes can have dynamic activation instantiations (e.g. if i recall correctly RNNs can have specific input activations functions)

**Comment by Adel-Moumen** [comment]
Hi guys @pplantinga @TParcollet @mravanelli , I think the PR is now ready. I went through the comments, and added a tutorial for the caching feature. Also, I improved the ASR SpeechLLM baseline and now gets more competitive results (2.72% on LS test-clean and 5.34% on test-other). 

I also created the model card and will upload the required files so that we can display this example for the community. This is based on Llama 3.2 1B with LORA adapters and WavLM-Large. 

Happy to consider other remarks/requests. 

:)

File: speechbrain/nnet/activations.py
Comment (pplantinga) [reply speechbrain/nnet/activations.py:1]: I get that sometimes a class may be needed and sometimes a function, but torch has both a class `torch.nn.GELU` and a function `torch.nn.functional.gelu` so this shouldn't be needed here right?

Comment (pplantinga) [reply speechbrain/nnet/activations.py:1]: If you need the default argument changed you can always use "name" to change it, e.g.

```yaml
activation: !name:torch.nn.GELU {approximate: tanh}
```

Which returns a constructor that will use tanh by default.

File: speechbrain/nnet/activations.py [hunk 8]
```diff
@@ -169,3 +169,76 @@ def forward(self, x):
...
+            Input tensor.
+
+        Returns
+        -------
+        The GeLU output.
+        """
+        return F.gelu(x, approximate=self.approximate)
+
+
+class ReLU(torch.nn.Module):
```
Comment (pplantinga) [inline speechbrain/nnet/activations.py]: Same here, I don't think this is needed.

File: speechbrain/nnet/schedulers.py [hunk 9]
```diff
@@ -300,7 +300,6 @@ class LinearWarmupScheduler:
     from the initial lr set in the optimizer to 0, after
     a warmup period during which it increases linearly
     from 0 to the initial lr set in the optimizer.
-    * Ge Li 2022
```
Comment (pplantinga) [inline speechbrain/nnet/schedulers.py:303]: Did Ge Li not work on this? Why would we remove it?

File: speechbrain/core.py [hunk 10]
```diff
@@ -457,6 +457,12 @@ def __init__(  # noqa: C901
                 self.hparams.output_folder,
             )
 
+        self.raw_modules = (
+            self.modules.module
+            if hasattr(self.modules, "module")
+            else self.modules
+        )
+
```
Comment (pplantinga) [inline speechbrain/core.py:465]: I'm not sure about this... is this handling the extra layer from DDP? I'm wondering if this should just be handled by the recipe or if we do need a more general solution whether we need something more robust here somehow, like a function you can call to get the modules appropriately.

File: docs/tutorials/basics/data-loading-pipeline.ipynb [hunk 11]
```diff
@@ -1,2167 +1,2691 @@
...
...Dr+Rf8q3aCxzFCrvcDegRrLKaNPZ9/mq281F9K5Hdo1pJfVUTAMgxLRWSvSrATcbNkglCjPJp7pYyz7O7IT7VBwYwF0fTPUljXD7R/mh4+KmVmVUfUhrw2nO6uUHl9ylLsU8ve2TcHhuNb4fcE/Cs+lRD+Te+wfl9E75iJIlOtlmin+D0fOKu85qy26j0tP3qkcOvsldqHKth2+/ggSLedQXLv3ECdsxrC2cq7KyyWoInyG+dYrsO92xuTNLzf26lXXoXsf0OTeUNkeSkFvtFveb8qy5rZw8/esy2aX6pboayPhwDiDzlfmygvVq/RyOnz3LQKF+uWtYZPtB/3mnhHDUXBjEbjp+kxI5jbHhThRu+XCaXH6z39RtUS2xEXluVrYECbO6J3ZbqO4CqxS/Bm5x9lvRo+clX8jLPxZ9TLc6LtH8WrTcJ3bNkm8AU/eaPdzf7Bffe4UXdkkZAWz8cczEv75J+s3YfzB9mmoIswKFIWMaeVj0UebPxZn7P1lAvN3s7+jJKfpnmuqvxX7e7jZvDmjddmHO/7UXIgFjIb/rLJp6nN8Ee5QcGMBuHovLJL2THMhA3xK0G6eBBc9U8pW4nAhOVX+NWbPEcMwDJLTsj6U812dpXW9To+3oZ+V8kRw7umR6LzyCm6+UT6/wzr2pdLjUikjtzLmNzG786GqCrKW6QalGDaMNE4sP3E66dMbpftxmZO9NhMRfvI4PsWpX/HGhhsaEj5m/4vbfdY4fysaSbWfVzVGzH3izcjYH/hvv/r8QgUE3ExkJppRcGMBuBqWKij4zEm9mU4e2KRlSeNOynvG4oZ3TJrqob2F+y6iVny47HFzNUvic8qZFTU7J8EPbPkYit0rMyZW5lweruynKZUy6DtvE35ZuBUMw+BFDLcrZITI6A1gNGy/oK16orvYfcB8V1FlKvz1IkKNEHg3fKV+q4z4+Kz5YvXfstuDp6/YRMOC1/mH2N3wc/+eLZgrVhx6JqaBghvCm8zhG02U5YrQdt6IPl5eVOyxSNWjax4AGEb1RMbGtwaoXZH15dllrDilf2K+zEncGzetkX9Cyc/z8tO3WJ04GKviBiAuIQkLly3S+7rKxP6Qv+uW/H9C9PcEw3ONZCr2KGOZ9ZH7kbj0gtvAXB/aDO+8iDHOTtI+GhIyZl99VEmo3V5sXIu+aVjw2vtac1a2c8lU6Vu45
```
Comment (pplantinga) [inline docs/tutorials/basics/data-loading-pipeline.ipynb]: I'm confused why this states "Compression: None" when "gzip" was specified earlier. Maybe this just needs to be run once more?

**Review by pplantinga — COMMENTED** [review]
Looks like its getting close, thanks Adel! Just a few small comments before merge.

File: speechbrain/core.py (hunk 10 above)
Comment (Adel-Moumen) [reply speechbrain/core.py:465]: The point is to exactly remove the DDP extra layering (similar to here: https://github.com/karpathy/nanoGPT/blob/3adf61e154c3fe3fca428ad6bc3818b27a3b8291/train.py#L253). That way, we can systematically access to methods e.g. `get_input_embedding` which wouldn't be accessible easily (e.g. here: `self.raw_modules.get_input_embedding`) while before we would have to do something like `x = self.modules.module.get_input_embedding if 'module' in self.modules else ...` so you would still need to unwrap the DDP container to get target module for calling the function. Here, we just have a pointer that does that instead of doing it manually. I have seen some recipes that had to do a ugly 'if' and I think this would solve the problem. 

having a function that unwrap would work but I think this is simpler and more modular. Happy to consider something else if you think this is not the way to go.

**Comment by Adel-Moumen** [comment]
> Looks like its getting close, thanks Adel! Just a few small comments before merge.

I think I addressed all the comments! Fixed the notebook and HDF5 (we were never passing the `compression` arg to the constructor) + removed the activations + added back authors.

File: speechbrain/core.py (hunk 10 above)
Comment (pplantinga) [reply speechbrain/core.py:465]: Well, I'm happy to go with this for now if needed, but maybe the longer term solution is to actually store the DDP module in a separate container. I mean for this part (simplified):

```
for name, module in self.modules.items():
    if any(p.requires_grad for p in module.parameters()):
        module = SyncBatchNorm.convert_sync_batchnorm(module)
        ddp_module = DDP( 
            module,
            device_ids=[self.device],
            find_unused_parameters=self.find_unused_parameters,
        )
        self.modules[name] = ddp_module
```

change the last line to `self.ddp_modules[name] = ddp_module` so we don't overwrite the old one.
